src/Widgets/CalendarChooser.vala

```diff
@@ -0,0 +1,275 @@
+            return _current_source;
+        }
+        set {
+            _current_source = value;
+            calendar_grid.source = value;
+            tooltip_text = "%s - %s".printf (calendar_grid.label, calendar_grid.location);
+        }
+    }
+
+    private Gtk.SearchEntry search_entry;
+    private CalendarGrid calendar_grid;
+
+    construct {
+        sources = new GLib.List<E.Source> ();
+        var calmodel = Calendar.EventStore.get_default ();
+        var registry = calmodel.registry;
+        foreach (var src in registry.list_sources (E.SOURCE_EXTENSION_CALENDAR)) {
+            if (src.writable == true && src.enabled == true && calmodel.calclient_is_readonly (src) == false) {
+                sources.append (src);
+            }
+        }
+
+        _current_source = registry.default_calendar;
+
+        current_source = registry.default_calendar;
+
+        search_entry = new Gtk.SearchEntry ();
+        search_entry.margin = 12;
+        search_entry.margin_bottom = 6;
+        search_entry.placeholder_text = _("Search Calendars");
+
+        var placeholder = new Granite.Widgets.AlertView (
+            _("No Results"),
+            _("Try changing search terms."),
+            ""
+        );
+        placeholder.show_all ();
+
+        var list_box = new Gtk.ListBox ();
+        list_box.activate_on_single_click = true;
+        list_box.set_placeholder (placeholder);
+
+        var list_bg_css = "list { background: transparent; }";
+        try {
+            var css_provider = new Gtk.CssProvider ();
+            css_provider.load_from_data (list_bg_css);
+            list_box.get_style_context ().add_provider (css_provider, Gtk.STYLE_PROVIDER_PRIORITY_APPLICATION);
+        } catch (Error e) {
+            warning ("Could not create CSS Provider: %s\nStylesheet:\n%s", e.message, list_bg_css);
+        }
+
+        var scrolled = new Gtk.ScrolledWindow (null, null);
+        scrolled.hscrollbar_policy = Gtk.PolicyType.NEVER;
+        scrolled.add (list_box);
+        scrolled.max_content_height = 300;
+        scrolled.propagate_natural_height = true;
+
+        margin_bottom = 6;
+        attach (search_entry, 0, 0);
+        attach (scrolled, 0, 1);
+        show_all ();
+
+        list_box.set_filter_func (filter_function);
+        list_box.set_header_func (header_update_func);
+
+        list_box.set_sort_func ((row1, row2) => {
+            var child1 = (CalendarGrid)row1.get_child ();
+            var child2 = (CalendarGrid)row2.get_child ();
+            var comparison = child1.location.collate (child2.location);
+            if (comparison == 0) {
+                return child1.label.collate (child2.label);
+            } else {
+                return comparison;
+            }
+        });
+
+        list_box.move_cursor.connect ((mvmt, count) => {
+            var row = list_box.get_selected_row ().get_index ();
+            if (row == 0 && count == -1) {
+                search_entry.grab_focus ();
```

Comment:
Wouldn't this be more fragile, since it requires that the focus order never changes? If we always want to focus the search entry, it makes more sense to me to have it actively grab focus. But this isn't my code so I don't have a real connection to it 🙂️

Reply:
Well, theoretically, if say the search entry moved below the list for some reason, focusing the search entry when you move up would no longer be correct. But using `move_focus` whatever widget was there would be selected correctly. So it's more future proof, imo. But realistically this problem isn't an actual problem that will occur. Just minor nitpicking